src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+lemma is_nontrivial.comp {χ : mul_char R R'} (hχ : χ.is_nontrivial)
+  {f : R' →+* R''} (hf : function.injective f) :
+   (χ.ring_hom_comp f).is_nontrivial :=
+begin
+  obtain ⟨a, ha⟩ := hχ,
+  use a,
+  rw [ring_hom_comp_apply, ← ring_hom.map_one f],
+  exact λ h, ha (hf h),
+end
+
+/-- Composition with a ring homomorphism preserves the property of being a quadratic character. -/
+lemma is_quadratic.comp {χ : mul_char R R'} (hχ : χ.is_quadratic) (f : R' →+* R'') :
+  (χ.ring_hom_comp f).is_quadratic :=
+begin
+  intro a,
+  simp only [ring_hom_comp_apply],
+  rcases hχ a with (ha | ha | ha); rw [ha],
+  { left,
+    rw [ring_hom.map_zero], },
+  { right, left,
+    rw [ring_hom.map_one], },
+  { right, right,
+    rw [ring_hom.map_neg, ring_hom.map_one], },
+end
+
+/-- The inverse of a quadratic character is itself. →  -/
+lemma quadratic_char_inv {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ⁻¹ = χ :=
+begin
+  ext x,
+  rw [inv_apply_eq_inv],
+  rcases hχ x with h₀ | h₁ | h₂,
+  { rw [h₀, ring.inverse_zero], },
+  { rw [h₁, ring.inverse_one], },
+  { rw [h₂, (by norm_num : (-1 : R') = (-1 : R'ˣ)), ring.inverse_unit (-1 : R'ˣ)],
+    refl, },
+end
+
+/-- The square of a quadratic character is the trivial character. -/
+lemma quad_char_sq_eq_one {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ ^ 2 = 1 :=
+begin
+  rw [pow_two],
+  nth_rewrite 0 ← quadratic_char_inv hχ,
+  rw [mul_left_inv],
+end
+
+/-- The `p`th power of a quadratic character is itself, when `p` is the (prime) characteristic
+of the target ring. -/
+lemma quad_char_pow_char {χ : mul_char R R'} (hχ : χ.is_quadratic)
+ (p : ℕ) [hp : fact p.prime] [char_p R' p] :
+  χ ^ p = χ :=
+begin
+  ext x,
+  rw [pow_apply_coe],
+  rcases hχ x with (hx | hx | hx); rw hx,
+  { rw [zero_pow (fact.out p.prime).pos], },
+  { rw [one_pow], },
+  { exact char_p.neg_one_pow_char R' p, },
+end
+
+/-- The `n`th power of a quadratic character is itself, when `n` is odd. -/
+lemma quad_char_pow_odd {χ : mul_char R R'} (hχ : χ.is_quadratic) {n : ℕ} (hn : n % 2 = 1) :
+  χ ^ n = χ :=
+by rw [← nat.div_add_mod n 2, hn,
+       pow_add, pow_mul, quad_char_sq_eq_one hχ, one_pow, pow_one, mul_char.one_mul]
+
+open_locale big_operators
+
+/-- The sum over all values of a nontrivial multiplicative character on a finite ring is zero
+(when the target is a domain). -/
+lemma sum_eq_zero_of_is_nontrivial [fintype R] [is_domain R'] {χ : mul_char R R'}
+ (hχ : χ.is_nontrivial) :
+  ∑ a, χ a = 0 :=
+begin
+  rcases hχ with ⟨b, hb⟩,
+  have h₁ : ∑ a, χ (b * a) = ∑ a, χ a :=
+    fintype.sum_bijective _ (units.mul_left_bijective b) _ _ (λ x, rfl),
+  simp only [map_mul] at h₁,
+  rw [← finset.mul_sum] at h₁,
+  exact eq_zero_of_mul_eq_self_left hb h₁,
+end
+
+/-- The sum over all values of the trivial multiplicative character on a finite ring is
+the cardinality of its unit group. -/
+lemma sum_eq_card_units_of_is_trivial [fintype R] [decidable_eq R] :
```

Comment:
Since there's not really a definition of `is_trivial`, I'd name this after the terms appearing in the statement:
```suggestion
lemma sum_one_eq_card_units [fintype R] [decidable_eq R] :
```

Reply:
OK.